File: crates/pray-core/src/resolve.rs
use crate::lockfile::Lockfile;
use crate::manifest::{Manifest, ManifestPackage, ManifestSource};
use crate::package_spec::{parse_package_spec, PackageSpec};
use crate::resolve_context::ResolveOptions;
use crate::resolve_exports::{
    build_skill_file_index, load_export_bodies, load_package_file_bytes, select_exports,
};
use crate::resolve_git_sources::{prepare_git_sources, prepare_pray_ssh_host_keys, GitSourceSet};

use crate::paths::find_prayspec_file;
pub use crate::resolve_git::{discover_distribution_root, git_source_cache_directory};
pub use crate::resolve_git_lookup::git_source_cached_repository;
pub use crate::resolve_git_paths::git_source_cache_directory_with_subdir;
pub use crate::resolve_git_refresh::{
    annotate_failed_git_refresh, annotate_missing_git_catalog,
    resolution_may_benefit_from_git_source_refresh,
};
pub use crate::resolve_git_sources::refresh_git_sources;
use crate::{PrayError, PrayResult};
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone)]
pub struct ResolvedProject {
    pub manifest_path: PathBuf,
    pub project_root: PathBuf,
    pub manifest: Manifest,
    pub manifest_hash: String,
    pub packages: Vec<ResolvedPackage>,
    pub local_files: Vec<ResolvedLocalFile>,
    pub source_revisions: BTreeMap<String, String>,
    pub source_host_keys: BTreeMap<String, String>,
    pub environment: Option<String>,
}

#[derive(Debug, Clone)]
pub struct ResolvedPackage {
    pub declaration: ManifestPackage,
    pub root: PathBuf,
    pub spec: PackageSpec,
    pub tree_hash: String,
    pub artifact_hash: String,
    pub artifact: String,
    pub selected_exports: Vec<String>,
    pub source_checksum: String,
    pub export_bodies: BTreeMap<String, String>,
    pub skill_files: BTreeMap<String, Vec<String>>,
    pub signer_fingerprint: Option<String>,
    /// Highest non-yanked version in registry metadata when the package came from a registry source.
    pub registry_latest_version: Option<String>,
    /// True when the package was declared in Prayfile; false for transitive dependencies.
    pub explicit: bool,
    pub upstream: Option<crate::package_upstream::LockedUpstream>,
}

#[derive(Debug, Clone)]
pub struct ResolvedLocalFile {
    pub path: PathBuf,
    pub manifest_path: String,
    pub content: String,
    pub source_checksum: String,
    pub position: String,
    pub optional: bool,
}

impl ResolvedProject {
    pub fn lockfile_hash(&self) -> PrayResult<String> {
        Ok(self.manifest_hash.clone())
    }
}

pub fn project_root_from_manifest(manifest_path: &Path) -> PathBuf {
    match manifest_path.parent() {
        Some(parent) if !parent.as_os_str().is_empty() => parent.to_path_buf(),
        _ => PathBuf::from("."),
    }
}

fn canonical_project_root(manifest_path: &Path) -> PrayResult<PathBuf> {
    let root = project_root_from_manifest(manifest_path);
    if root.is_absolute() {
        return Ok(root);
    }
    let cwd = std::env::current_dir().map_err(|error| {
        PrayError::Resolution(format!("failed to resolve project root from cwd: {error}"))
    })?;
    Ok(cwd.join(root))
}

pub fn resolve_project(manifest_path: &Path) -> PrayResult<ResolvedProject> {
    resolve_project_with_options(manifest_path, &ResolveOptions::default())
}

pub fn resolve_project_with_git_refresh_fallback(
    manifest_path: &Path,
    options: &ResolveOptions,
    allow_git_refresh_fallback: bool,
) -> PrayResult<ResolvedProject> {
    match resolve_project_with_options(manifest_path, options) {
        Ok(project) => Ok(project),
        Err(PrayError::Resolution(message))
            if allow_git_refresh_fallback
                && !options.offline
                && !options.refresh_source_revisions
                && resolution_may_benefit_from_git_source_refresh(&message) =>
        {
            let refreshed_options = ResolveOptions {
                refresh_source_revisions: true,
                ..options.clone()
            };
            match resolve_project_with_options(manifest_path, &refreshed_options) {
                Ok(project) => Ok(project),
                Err(error) => {
                    let lockfile_path =
                        project_root_from_manifest(manifest_path).join("Prayfile.lock");
                    Err(annotate_failed_git_refresh(&lockfile_path, error))
                }
            }
        }
        Err(error) => Err(error),
    }
}

pub fn resolve_project_with_options(
    manifest_path: &Path,
    options: &ResolveOptions,
) -> PrayResult<ResolvedProject> {
    let project_root = canonical_project_root(manifest_path)?;
    resolve_project_in_context(manifest_path, &project_root, options)
}

#[path = "resolve_project.rs"]
mod project;
pub use project::{resolve_manifest_in_context, resolve_project_in_context};

#[path = "resolve_implied_source.rs"]
mod implied_source;
pub(crate) use implied_source::implied_source_name;

#[path = "resolve_package_root.rs"]
mod package_root;
pub(crate) use package_root::{resolve_package_root, PackageRootResolution};

#[path = "resolve_upstream.rs"]
pub mod resolve_upstream;
pub use resolve_upstream::{apply_path_upstream_refreshes, path_fork_drift_lines};

fn resolve_package(
    project_root: &Path,
    sources: &BTreeMap<String, ManifestSource>,
    git_sources: &GitSourceSet,
    user_config: &crate::config::PrayConfig,
    declaration: &ManifestPackage,
    lockfile: Option<&Lockfile>,
    options: &ResolveOptions,
) -> PrayResult<ResolvedPackage> {
    let PackageRootResolution {
        root,
        signer_fingerprint,
        registry_latest_version,
    } = resolve_package_root(
        project_root,
        sources,
        git_sources,
        user_config,
        declaration,
        lockfile,
        options,
    )?;
    let spec_path = find_prayspec_file(&root)?;
    let spec_text = fs::read_to_string(&spec_path)?;
    let spec = parse_package_spec(&spec_text)?.canonicalized();
    if spec.name != declaration.name {
        return Err(PrayError::Resolution(format!(
            "package path {:?} declares {:?}, expected {:?}",
            root, spec.name, declaration.name
        )));
    }
    spec.satisfy_constraint(&declaration.constraint)?;
    let selected_exports = select_exports(declaration, &spec)?;
    let file_bytes = load_package_file_bytes(&root, &spec)?;
    let tree_hash = PackageSpec::tree_hash_from_file_bytes(&file_bytes)?;
    let export_bodies = load_export_bodies(&file_bytes, &spec, &selected_exports)?;
    let skill_files = build_skill_file_index(&spec);
    let source_checksum = tree_hash.clone();
    let upstream_context = resolve_upstream::UpstreamResolutionContext::new(
        project_root,
        sources,
        git_sources,
        user_config,
        lockfile,
        options,
    );
    let upstream = resolve_upstream::lock_path_upstream(&upstream_context, declaration, &spec)?;
    Ok(ResolvedPackage {
        declaration: declaration.clone(),
        root,
        spec: spec.clone(),
        tree_hash: tree_hash.clone(),
        artifact_hash: tree_hash.clone(),
        artifact: format!(
            "path:{}",
            spec_path.parent().unwrap_or(&spec_path).to_string_lossy()
        ),
        selected_exports,
        source_checksum,
        export_bodies,
        skill_files,
        signer_fingerprint,
        registry_latest_version,
        explicit: false,
        upstream,
    })
}

#[path = "resolve_local.rs"]
mod local;
pub use local::missing_local_embed_guidance;
pub(crate) use local::resolve_local_file;

fn source_map(sources: &[ManifestSource]) -> BTreeMap<String, ManifestSource> {
    sources
        .iter()
        .map(|source| (source.name.clone(), source.clone()))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::{discover_distribution_root, project_root_from_manifest};
    use std::fs;
    use std::path::Path;

    #[test]
    fn project_root_from_manifest_uses_cwd_for_bare_filename() {
        let root = project_root_from_manifest(Path::new("Prayfile"));
        assert_eq!(root, Path::new("."));
    }

    #[test]
    fn project_root_from_manifest_uses_parent_directory() {
        let root = project_root_from_manifest(Path::new("examples/simple-project/Prayfile"));
        assert_eq!(root, Path::new("examples/simple-project"));
    }

    #[test]
    fn discover_distribution_root_finds_root_and_prayers_subdirectory() {
        let workspace =
            std::env::temp_dir().join(format!("pray-discover-distribution-{}", std::process::id()));
        let _ = fs::remove_dir_all(&workspace);
        let repo_root = workspace.join("repo");
        let prayers_root = repo_root.join("prayers");
        fs::create_dir_all(prayers_root.join("v1/packages")).expect("prayers distribution");
        fs::create_dir_all(repo_root.join("v1/packages")).expect("root distribution");

        assert_eq!(
            discover_distribution_root(&repo_root),
            Some(repo_root.clone())
        );

        fs::remove_dir_all(repo_root.join("v1")).expect("remove root distribution");
        assert_eq!(discover_distribution_root(&repo_root), Some(prayers_root));
        let _ = fs::remove_dir_all(&workspace);
    }

    #[test]
    fn discover_distribution_root_returns_none_without_registry_layout() {
        let workspace =
            std::env::temp_dir().join(format!("pray-discover-missing-{}", std::process::id()));
        let _ = fs::remove_dir_all(&workspace);
        let repo_root = workspace.join("repo");
        fs::create_dir_all(&repo_root).expect("repo root");
        assert_eq!(discover_distribution_root(&repo_root), None);
        let _ = fs::remove_dir_all(&workspace);
    }
}

File: rubygems/pray-cli/lib/pray/resolve.rb
# frozen_string_literal: true

module Pray
  ResolvedProject = Struct.new(
    :manifest_path, :project_root, :manifest, :manifest_hash, :packages,
    :local_files, :source_revisions, :source_host_keys,
    keyword_init: true
  ) do
    def lockfile_hash
      manifest_hash
    end
  end

  ResolvedPackage = Struct.new(
    :declaration, :root, :spec, :tree_hash, :artifact_hash, :artifact,
    :selected_exports, :source_checksum, :export_bodies, :skill_files,
    :signer_fingerprint, :registry_latest_version,
    keyword_init: true
  )

  ResolvedLocalFile = Struct.new(
    :path, :manifest_path, :content, :position, :optional,
    keyword_init: true
  )

  module Resolve
    module_function

    def project_root_from_manifest(manifest_path)
      parent = File.dirname(manifest_path)
      parent.empty? || parent == "." ? "." : parent
    end

    def canonical_project_root(manifest_path)
      root = project_root_from_manifest(manifest_path)
      return Pathname(root).cleanpath.to_s if Pathname(root).absolute?

      File.expand_path(root)
    end

    def resolve_project(manifest_path, offline: false, refresh: false)
      user_config = Config.load_user_config
      project_root = canonical_project_root(manifest_path)
      lockfile_path = File.join(project_root, "Prayfile.lock")
      lockfile_hints = File.exist?(lockfile_path) ? Pray.read_lockfile(lockfile_path) : nil
      manifest_text = Pray.read_manifest_text(manifest_path)
      manifest = Pray.parse_manifest(manifest_text)
      manifest_hash = manifest.manifest_hash
      sources = source_map(manifest.sources)
      git_sources = GitSources.prepare_git_sources(
        project_root,
        manifest.sources,
        lockfile_hints,
        refresh: refresh
      )
      source_revisions = git_sources.transform_values(&:revision)
      source_host_keys = Trust.prepare_source_host_keys(manifest.sources)

      packages = []
      seen = {}
      errors = []
      manifest.packages.each do |declaration|
        begin
          package = resolve_package(
            project_root,
            sources,
            git_sources,
            user_config,
            declaration,
            lockfile_hints,
            offline: offline
          )
          if seen[package.declaration.name]
            raise Error.resolution("duplicate package declaration: #{package.declaration.name}")
          end

          seen[package.declaration.name] = true
          packages << package
        rescue Error => error
          errors << "#{declaration.name}: #{error.message}"
        end
      end
      raise Error.resolution(errors.join("\n")) unless errors.empty?

      local_files = []
      local_errors = []
      manifest.local.each do |local|
        begin
          local_files << resolve_local_file(project_root, local)
        rescue Error => error
          local_errors << "local #{local.path}: #{error.message}"
        end
      end
      raise Error.resolution(local_errors.join("\n")) unless local_errors.empty?

      ResolvedProject.new(
        manifest_path: manifest_path,
        project_root: project_root,
        manifest: manifest,
        manifest_hash: manifest_hash,
        packages: packages,
        local_files: local_files,
        source_revisions: source_revisions,
        source_host_keys: source_host_keys
      )
    end

    def missing_local_embed_guidance(path)
      "Prayfile lists `local \"#{path}\"` but the file does not exist. " \
        "Create the file or remove the entry from Prayfile, then run `pray install`."
    end

    def resolve_package(project_root, sources, git_sources, user_config, declaration, lockfile, offline: false)
      root, registry_latest_version = resolve_package_root_with_metadata(
        project_root, sources, git_sources, user_config, declaration, lockfile, offline: offline
      )
      spec_path = find_prayspec_file(root)
      spec_text = File.read(spec_path)
      spec = Pray.parse_package_spec(spec_text).canonicalized
      if spec.name != declaration.name
        raise Error.resolution(
          "package path #{root.inspect} declares #{spec.name.inspect}, expected #{declaration.name.inspect}"
        )
      end
      unless Constraint.version_satisfies(spec.version, declaration.constraint)
        raise Error.resolution(
          "package #{declaration.name} version #{spec.version} does not satisfy constraint #{declaration.constraint}"
        )
      end

      selected_exports = select_exports(declaration, spec)
      file_bytes = load_package_file_bytes(root, spec)
      tree_hash = PackageSpec.tree_hash_from_file_bytes(file_bytes)
      export_bodies = load_export_bodies(file_bytes, spec, selected_exports)
      skill_files = build_skill_file_index(spec)

      ResolvedPackage.new(
        declaration: declaration,
        root: root,
        spec: spec,
        tree_hash: tree_hash,
        artifact_hash: tree_hash,
        artifact: "path:#{File.dirname(spec_path)}",
        selected_exports: selected_exports,
        source_checksum: tree_hash,
        export_bodies: export_bodies,
        skill_files: skill_files,
        signer_fingerprint: nil,
        registry_latest_version: registry_latest_version
      )
    end

    def resolve_package_root_with_metadata(
      project_root, sources, git_sources, user_config, declaration, lockfile, offline: false
    )
      if (local_path = user_config.local.package[declaration.name])
        return [File.expand_path(local_path, project_root), nil]
      end
      return [File.expand_path(declaration.path, project_root), nil] if declaration.path

      if declaration.source
        source = sources[declaration.source]
        raise Error.resolution("unknown source: #{declaration.source}") unless source

        if (local_path = user_config.local.source[declaration.source])
          source_root = File.expand_path(local_path, project_root)
          resolved = Registry.resolve_local_registry_package_root(
            project_root,
            "local:#{declaration.source}",
            source_root,
            declaration,
            preferred_version: lockfile_preferred_version(lockfile, declaration.name),
            offline: offline
          )
          return [resolved.root, resolved.registry_latest_version]
        end

        case source.kind
        when "path"
          return [File.join(project_root, source.url, declaration.name.tr("/", "-")), nil]
        when "registry", "static index"
          resolved = Registry.resolve_registry_package_root(
            project_root,
            source.url,
            declaration,
            preferred_version: lockfile_preferred_version(lockfile, declaration.name),
            offline: offline
          )
          return [resolved.root, resolved.registry_latest_version]
        when "pray_ssh"
          raise Error.unsupported("pray_ssh sources are not implemented yet in pray-cli Ruby")
        when "git"
          checkout = git_sources[source.name]
          raise Error.resolution("git source #{source.name} was not prepared") unless checkout

          clone_url = source.url.delete_prefix("git+")
          distribution_root = GitSources.resolve_distribution_root(checkout.cache_directory, checkout.subdir)
          source_key = checkout.revision.to_s.empty? ? clone_url : "#{clone_url}@#{checkout.revision}"
          resolved = Registry.resolve_local_registry_package_root(
            project_root,
            source_key,
            distribution_root,
            declaration,
            preferred_version: lockfile_preferred_version(lockfile, declaration.name),
            offline: offline
          )
          return [resolved.root, resolved.registry_latest_version]
        else
          raise Error.unsupported("source kind #{source.kind} not implemented yet")
        end
      end

      if declaration.git || declaration.tarball || declaration.oci
        raise Error.unsupported("remote sources are not implemented yet")
      end

      [File.join(project_root, declaration.name.tr("/", "-")), nil]
    end

    def resolve_package_root(project_root, sources, git_sources, user_config, declaration, lockfile = nil)
      resolve_package_root_with_metadata(project_root, sources, git_sources, user_config, declaration, lockfile).first
    end

    def lockfile_preferred_version(lockfile, package_name)
      return nil unless lockfile

      lockfile.package.find { |entry| entry.name == package_name }&.version
    end

    def resolve_local_file(project_root, declaration)
      path = File.join(project_root, declaration.path)
      unless File.exist?(path)
        if declaration.optional
          return ResolvedLocalFile.new(
            path: path,
            manifest_path: declaration.path,
            content: "",
            position: declaration.position,
            optional: true
          )
        end
        raise Error.resolution(missing_local_embed_guidance(declaration.path))
      end

      ResolvedLocalFile.new(
        path: path,
        manifest_path: declaration.path,
        content: Hashing.normalize_line_endings(File.read(path)),
        position: declaration.position,
        optional: declaration.optional
      )
    end

    def find_prayspec_file(root)
      files = Dir.children(root).filter_map do |entry|
        path = File.join(root, entry)
        File.file?(path) && File.extname(entry) == ".prayspec" ? path : nil
      end
      case files.length
      when 1 then files.first
      when 0 then raise Error.resolution("no prayspec file found in #{root.inspect}")
      else raise Error.resolution("multiple prayspec files found in #{root.inspect}")
      end
    end

    def source_map(sources)
      sources.to_h { |source| [source.name, source] }
    end

    def select_exports(declaration, spec)
      return spec.exports.keys.sort if declaration.exports.empty?

      declaration.exports.each do |export|
        unless spec.exports.key?(export)
          raise Error.resolution("package #{declaration.name} does not export #{export}")
        end
      end
      declaration.exports
    end

    def load_package_file_bytes(root, spec)
      file_bytes = {}
      spec.files.each do |file|
        path = File.join(root, file)
        raise Error.integrity("package file missing: #{file}") unless File.exist?(path)
        raise Error.integrity("package file is a directory: #{file}") if File.directory?(path)

        file_bytes[file] = File.binread(path)
      end
      file_bytes
    end

    def load_export_bodies(file_bytes, spec, selected_exports)
      export_bodies = {}
      selected_exports.each do |export_name|
        entry = spec.exports[export_name]
        raise Error.resolution("package #{spec.name} is missing export #{export_name}") unless entry
        next unless entry.kind == "fragment"

        bytes = file_bytes[entry.path]
        raise Error.integrity("package file missing for export #{export_name}: #{entry.path}") unless bytes

        export_bodies[export_name] = Hashing.normalize_line_endings(bytes.force_encoding(Encoding::UTF_8))
      end
      export_bodies
    end

    def build_skill_file_index(spec)
      index = {}
      spec.exports.each do |export_name, export|
        next unless %w[folder skill].include?(export.kind)

        folder_prefix = export.path.delete_suffix("/")
        files = indexed_files_under_prefix(spec.files, folder_prefix)
        index[export_name] = files unless files.empty?
      end
      spec.skills.each do |skill_name, skill|
        next if index.key?(skill_name)

        skill_prefix = skill.path.delete_suffix("/")
        files = indexed_files_under_prefix(spec.files, skill_prefix)
        index[skill_name] = files unless files.empty?
      end
      index
    end

    def indexed_files_under_prefix(files, prefix)
      files.filter_map { |file| skill_relative_file(file, prefix) }
    end

    def skill_relative_file(file, skill_prefix)
      return nil unless file.start_with?(skill_prefix)

      relative = file.delete_prefix(skill_prefix).delete_prefix("/")
      return nil if relative.empty? || file == skill_prefix

      relative
    end
  end
end
